pallets/xcm-benchmarks/src/generic/benchmarking.rs

```diff
@@ -0,0 +1,318 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use crate::{new_executor, XcmCallOf};
+use codec::Encode;
+use frame_benchmarking::{benchmarks, BenchmarkError};
+use frame_support::dispatch::GetDispatchInfo;
+use sp_std::vec;
+use xcm::{
+	latest::{prelude::*, MultiAssets},
+	DoubleEncoded,
+};
+
+benchmarks! {
+	query_holding {
+		let holding = T::worst_case_holding();
+
+		let mut executor = new_executor::<T>(Default::default());
+		executor.holding = holding.clone().into();
+
+		let instruction = Instruction::<XcmCallOf<T>>::QueryHolding {
+			query_id: Default::default(),
+			dest: T::valid_destination()?,
+			// Worst case is looking through all holdings for every asset explicitly.
+			assets: Definite(holding),
+			max_response_weight: u64::MAX,
+		};
+
+		let xcm = Xcm(vec![instruction]);
+
+	} : {
+		executor.execute(xcm)?;
+	} verify {
+		// The completion of execution above is enough to validate this is completed.
+	}
+
+	// This benchmark does not use any additional orders or instructions. This should be managed
+	// by the `deep` and `shallow` implementation.
+	buy_execution {
+		let holding = T::worst_case_holding().into();
+
+		let mut executor = new_executor::<T>(Default::default());
+		executor.holding = holding;
+
+		let fee_asset = Concrete(Here.into());
+
+		let instruction = Instruction::<XcmCallOf<T>>::BuyExecution {
+			fees: (fee_asset, 100_000_000).into(), // should be something inside of holding
+			weight_limit: WeightLimit::Unlimited,
+		};
+
+		let xcm = Xcm(vec![instruction]);
+	} : {
+		executor.execute(xcm)?;
+	} verify {
+
+	}
+
+	// Worst case scenario for this benchmark is a large number of assets to
+	// filter through the reserve.
+	reserve_asset_deposited {
```

Comment:
I don't recall why `ReserveAssetDeposited` is replicated here, since it makes no sense for both generic and fungibles benchmarks to benchmark the same instruction. It looks to me that we only use the fungibles benchmark, so this can probably be removed, both here and upstream.

Reply:
afaik we aren't using either of them.

Reply:
I'll probably remove both of those in cumulus.

Reply:
Actually no, let's leave the fungible stuff just in case.